library/core/src/hint.rs

```diff
@@ -98,9 +98,13 @@ use crate::intrinsics;
 #[rustc_const_stable(feature = "const_unreachable_unchecked", since = "1.57.0")]
 #[cfg_attr(miri, track_caller)] // even without panics, this helps for Miri backtraces
 pub const unsafe fn unreachable_unchecked() -> ! {
-    // SAFETY: the safety contract for `intrinsics::unreachable` must
-    // be upheld by the caller.
-    unsafe { intrinsics::unreachable() }
+    if cfg!(debug_assertions) {
```

Comment:
Pondering: is this going to work in user code?  I assume no, since we always ship core in non-debug.

I don't know if it's a good idea, but we could consider doing something like
```rust
let bomb = AbortOnPanic::new();
let _ = u32::sub(0, 1);
bomb.defuse();
unsafe { intrinsics::unreachable() }
```

so it'd work in shipped core too.

Reply:
> Pondering: is this going to work in user code? I assume no, since we always ship core in non-debug.

No. I assumed the idea was that this would only affect std when std itself is build in debug mode, like with -Zbuild-std and in some of rust-lang/rust's CI tests. If this is meant to also impact regular debug builds (for which we use a non-debug std), then we should probably have some more discussion before accepting a change like this.